client.go

```diff
@@ -0,0 +1,77 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package cluster
+
+// KVStore provides access to the configuration store
+type KVStore interface {
+	// Get retrieves the value for the given key, marshalling it into the provided value
+	Get(key string, v interface{}) error
+
+	// Put stores the value for the given key
+	Put(key string, v interface{}) error
+}
+
+// A ServiceInstance is a single instance of a service
+type ServiceInstance struct {
```

Comment:
That said, I'd dispute that everything being interfaces makes testing less painful - now you need to mock non-behavioral value objects, where as if they were just structs you could just use them.

FWIW our mocking approach is really unmaintainable.  We need a deterministic build step that doesn't require developers to know exactly what mocks exists and how we are mapping interfaces into mock files.